### ISSUE TITLE
Add support for restoring to another zone [GCP]

### ISSUE THREAD

**Issue by kanuahs** [vmware-tanzu/velero#1624]
**Describe the solution you'd like**
I've been evaluating velero for disaster recovery. One of the use cases is restoring a failed cluster in another availability zone. I want to take a backup of a GKE cluster which is running in one zone (eg. us-east1-b) and restore it in another zone in the same region (eg. us-east1-c).
Apart from Statefulsets & PVs, everything seems to be working.
The PVs get provisioned in the first zone (us-east1-b) and can't be mounted.

When restoring PVs from snapshots, the AZ of the previous volume is used. I'd like to overrride this when restoring the backup.

**Anything else you would like to add:**
Persistent disk snapshots in GCP are regional or multi-regional, so it is possible to restore them into another zone.

**Environment:**

- Velero version (use `velero version`):
Client:
	Version: v1.0.0
	Git commit: 72f5cadc3a865019ab9dc043d4952c9bfd5f2ecb
Server:
	Version: v1.0.0

- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"clean", BuildDate:"2019-05-16T16:23:09Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.10-gke.5", GitCommit:"5aa3a95d828fe45aab3611dfc4ebdc0341fe1507", GitTreeState:"clean", BuildDate:"2019-05-29T17:25:39Z", GoVersion:"go1.10.8b4", Compiler:"gc", Platform:"linux/amd64"}

- Kubernetes installer & version: GKE
- Cloud provider or hardware configuration: 4x n1-standard-4
- OS (e.g. from `/etc/os-release`): Container-Optimized OS

**Comment by skriss** [comment]
Thanks for the request @kanuahs! We'll evaluate and update with more detail/questions here.

**Comment by nrb** [comment]
I believe this is already possible using [regional disks in Kubernetes](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/regional-pd). PR #1200 fixed a bug that prevented it, but it should work as long as a regional disk is used.

**Comment by dijitali** [comment]
@nrb - unfortunately regional disks have a minimum size of 200GB:

> Regional persistent disk limits are similar to zonal persistent disks. However, regional standard persistent disks have a 200 GB size minimum.

From [Regional persistent disks (Standard and SSD): Limits](https://cloud.google.com/compute/docs/disks/#repds)

This puts the cheapest EU disk at $68/month each (using `europe-west1`).  This adds a significant cost to storage when smaller disks would suffice (we currently use several PersistentVolumes with <20GB disks, costing $3.4/month).

I realise this is a restriction incurred by GCP, not Velero, but as a result it makes this solution a bit untenable for our use case.

It sounds like this might be addresses by the work in #103 though?

**Comment by nrb** [comment]
That makes a whole lot of sense to me, @dijitali.

Yes, addressing #103 is what would enable this from the Velero side. It was in the planning stages but was de-prioritized. The `BackupStorageLocation` and `VolumeSnapshotLocation` CRDs were introduced as prerequisites for the work.

@VMmore and @skriss, we should revisit where we think replication should go in the long term roadmap. I still believe it's likely to a major release.

**Comment by skriss** [comment]
xref #1755

**Comment by kanuahs** [comment]
GCP PersistentDisk snapshots are regional(us-east) or multi-regional(us) even for zonal PersistentDisks.
They can be restored into another zone without doing any replication.

The node affinity field in PersistentVolume templates is causing the restored GCP PersistentDisks to be provisioned in the wrong zone (the source cluster's zone).

I think this issue could be fixed for GCP by either removing the node affinity and letting it get auto-generated when restoring or adding an option to override it with another zone/region before restoring PVs.

I was able to get velero to restore into another zone by downloading the compressed archives from the bucket and changing the NodeAffinity field in PersistentVolume templates from source cluster's zone (us-east1-b) to target cluster's zone (us-east1-c) as well as the volumeAZ field in the VolumeSnapshots.json file.

**Comment by mizunos** [comment]
I am experiencing this issue when trying to restore PVC to a cluster in another GCP project and region.  A couple of questions about the work around you mentioned about i.e NodeAffinity field in the PersistentVolume templates :1 -  is it being done to the backup zip files before the restore then restore? 2 - I can't seem to find any persistent volume template in the backup file (there are pvc objects that are backed up),  can you give more specific directory names>.  Thanks

**Comment by stevegore** [comment]
I'll share the script I created to resolve my own issue.

I needed to do this all in a Docker container as my Mac kept inserting `._` files.

1. Spin up a container with the GCP SDK: `docker run --rm -it google/cloud-sdk:alpine`
2. Authenticate: `gcloud auth login`
3. Run the below

```bash
export PROJECT_NAME=yourvalueshere
export BACKUP_NAME=yourvalueshere
export BUCKET_NAME=yourvalueshere
export OLD_REGION=yourvalueshere
export OLD_ZONE=yourvalueshere
export NEW_REGION=yourvalueshere
export NEW_ZONE=yourvalueshere

gcloud config set project "$PROJECT_NAME"

mkdir fix
cd fix

# Volume snapshots
gsutil cp gs://$BUCKET_NAME/backups/$BACKUP_NAME/$BACKUP_NAME-volumesnapshots.json.gz  .
gunzip $BACKUP_NAME-volumesnapshots.json.gz
sed -i "s/$OLD_ZONE/$NEW_ZONE/g" $BACKUP_NAME-volumesnapshots.json
sed -i "s/$OLD_REGION/$NEW_REGION/g" $BACKUP_NAME-volumesnapshots.json
gzip $BACKUP_NAME-volumesnapshots.json

# Bigger file
gsutil cp gs://$BUCKET_NAME/backups/$BACKUP_NAME/$BACKUP_NAME.tar.gz .
mkdir $BACKUP_NAME-temp
tar xzf $BACKUP_NAME.tar.gz -C $BACKUP_NAME-temp
cd $BACKUP_NAME-temp
find . -name \*.json -exec sh -c "sed -i 's/$OLD_ZONE/$NEW_ZONE/g' {}" \;
find . -name \*.json -exec sh -c "sed -i 's/$OLD_REGION/$NEW_REGION/g' {}" \;
tar czf ../$BACKUP_NAME.tar.gz *
gsutil cp ../$BACKUP_NAME.tar.gz gs://$BUCKET_NAME/backups/$BACKUP_NAME
```

**Comment by skriss** [comment]
Cool, thanks for that @stevegore!

**Comment by winmillwill** [comment]
Thanks for the script @stevegore -- I had to make some changes to get it to work for me, but knowing that it must have worked for you helped me persevere. Most notably, it seems like your version doesn't copy the gzipped volumesnapshots.json back to the bucket.

Everything else is just appeasing shellcheck, using bash "strict mode", and making the `sed -i` portable so it works on mac.

```
#!/bin/bash

set -eu -o pipefail

PROJECT_NAME=${PROJECT_NAME:-yourdefaultproject}
BACKUP_NAME=yourbackupname
BUCKET_NAME=yourbucketname
# this is what you'd do if the two kube clusters agreed on 2 of 3 AZs
# ... but one of them had AZ a and the other had AZ f
ZONE_MAP="a:f"
OLD_REGION="us-central1"
NEW_REGION="us-central1"

# sed_for_zone_map takes the old_region and new_region and an optional mapping of
# zones and returns a sed command to replace az and region occurrences as desired.
#
# the map is a string of the form "t:u,v:w,x:y" where the commas separate pairs
# and the pair is separated by a colon, and each pair maps occurrences of the
# first to the second. These are just the letter of the az. They DO NOT include
# the region to avoid the error case where the zone_map could be inconsistent
# with the old_region and new_region.
sed_for_zone_map() {
    local old_region=$1
    local new_region=$2
    local zone_map="${3:-}"
    local old_ifs="$IFS"
    local map_pairs=()
    local sed_commands=()

    IFS=","
    read -ra map_pairs <<< "$zone_map"
    IFS="$old_ifs"

    for pair in "${map_pairs[@]}"; do
        new_pair=()
        IFS=":"
        read -ra new_pair <<< "$pair"
        IFS="$old_ifs"
        sed_commands=("${sed_commands[@]}" "-e" "s/${old_region}-${new_pair[0]}/${new_region}-${new_pair[1]}/g")
    done

    # append a command to replace the region so that
    # 1. we always correct the az's first when the map is not empty
    # 2. we replace any remaining instance of $old_region with $new_region, either
    #   a. just a region identifier we need to update OR
    #   b. an identity mapped az that just needs its region updated
    sed_commands=("${sed_commands[@]}" "-e" "s/${old_region}/${new_region}/g")

    echo "sed -i.bak" "${sed_commands[@]}"
}

main() {
    gcloud config set project "$PROJECT_NAME"

    err=""
    dir=$(mktemp -d /tmp/fix-velero-${BACKUP_NAME}.XXXX) || err="could not create tempdir"
    if [[ -n $err ]]; then
        echo "$err"
        exit 1
    fi

    cd "$dir"

    # Volume snapshots
    gsutil cp gs://$BUCKET_NAME/backups/$BACKUP_NAME/$BACKUP_NAME-volumesnapshots.json.gz  .
    gunzip $BACKUP_NAME-volumesnapshots.json.gz
    $(sed_for_zone_map $OLD_REGION $NEW_REGION $ZONE_MAP) $BACKUP_NAME-volumesnapshots.json
    gzip $BACKUP_NAME-volumesnapshots.json
    gsutil cp ./$BACKUP_NAME-volumesnapshots.json.gz gs://$BUCKET_NAME/backups/$BACKUP_NAME/$BACKUP_NAME-volumesnapshots.json.gz

    # Bigger file
    gsutil cp gs://$BUCKET_NAME/backups/$BACKUP_NAME/$BACKUP_NAME.tar.gz .
    mkdir $BACKUP_NAME-temp
    tar xzf $BACKUP_NAME.tar.gz -C $BACKUP_NAME-temp
    cd $BACKUP_NAME-temp
    find . -name \*.json -exec sh -c "$(sed_for_zone_map $OLD_REGION $NEW_REGION $ZONE_MAP) \$1" _ {} \;
    tar czf ../$BACKUP_NAME.tar.gz ./*
    gsutil cp ../$BACKUP_NAME.tar.gz gs://$BUCKET_NAME/backups/$BACKUP_NAME
}

# comment this line out if you want to source this file and test the sed_for_zone_map function
main
```

**Comment by stevegore** [comment]
> Most notably, it seems like your version doesn't copy the gzipped volumesnapshots.json back to the bucket.

Looks like I made some mistakes in trying to make my script "GitHub ready" :). Regardless yours is way better - thank you!

**Comment by kmadel** [comment]
Velero added beta support for CSI with 1.4. By enabling the Kubernetes native support for VolumeSnapshots with CSI with Velero 1.4 you can now have multi-zone and multi-region restore for GCP/GKE as GCP disk snapshots are multi-region when created via the CSI Volume Snapshot capability. The only caveat that I am aware of is that GCP multi-region disk snapshots are still geo bound to: us, europe, or asia regions - so a recovery in europe from a backup in us may not work. However, I successfully tested creating a backup of a StatefulSet with a PV in us-central1 and successfully restored to us-east4 with no extra steps or manual modifications of the backup json.

**Comment by sreesanpd** [comment]
@kmadel I tried using CSI plugin with Velero 1.4.2 in gke cluster (k8s version 1.17) with the addon GcePersistentDiskCsiDriver=ENABLED. Velero took multi region snapshots but when I tried to restore it in another region, the PV got created in the primary region and not in the secondary region (so the pods couldn't bind the PV). It would be great if you can describe the exacts steps that you have followed to achieve this as I couldn't find any documents describing these steps for GKE using velero CSI plugin

The steps I followed are given in : https://stackoverflow.com/questions/63460096/unable-to-restore-gcp-persistent-disks-pvcs-to-another-region-when-backing-up-gk

**Comment by kmadel** [comment]
I have only tested with GCP Regional Persistent Disks and I created regions specific storage classes in the backup region and restore regions. Here is an example storage class that uses Regional Disk:

```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
  name: regional-pd-ssd-csi-storageclass
provisioner: pd.csi.storage.gke.io
parameters:
  type: pd-ssd
  #replication-type is one of none or regional-pd, defaults to none (zonal PD)
  replication-type: regional-pd
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
allowedTopologies:
- matchLabelExpressions:
  - key: topology.gke.io/zone
    values:
    - us-central1-a
    - us-central1-b
```
I posted more details in: https://stackoverflow.com/a/63504435/1031472

**Comment by dsu-igeek** [comment]
Merged into #3772
